### PR TITLE
workflows: dismiss PR approvals on non-maintainer pushes

### PR DESCRIPTION
**Pull request by ZhongRuoyu**
We'd like our branch protection rules to dismiss PR approvals when new commits are pushed by a user without write access to the repository. Unfortunately, GitHub's branch protection rules does not provide enough granularity to achieve that. So, let's do it with a workflow.

### CODE REVIEW

File: .github/workflows/dismiss-approvals.yml
Comment (carlocab) [inline .github/workflows/dismiss-approvals.yml:29]: IIRC using `!` is finicky when not wrapped with `${{`, `}}`. But also wrapping with `"`, `"` works, so maybe `(`, `)` does it too.

File: .github/workflows/dismiss-approvals.yml
Comment (carlocab) [inline .github/workflows/dismiss-approvals.yml:23]:
```suggestion
              /orgs/Homebrew/teams/core/members \
              --jq '[.[] | .login]'
```
If you only need one `jq` query then `gh` can handle it.

**Review by carlocab — DISMISSED** [review]

**Comment by ZhongRuoyu** [comment]
https://github.com/Homebrew/homebrew-core/actions/runs/4651866074/jobs/8231732430?pr=127939#step:2:23

```
Error: Unable to process file command 'output' successfully.
Error: Invalid format '  "alebcay", "bayandin", "bevanjkay", "Bo98", "branchvincent",'
```

Okay, can't put the JSON on multiple lines.

File: .github/workflows/dismiss-approvals.yml
Comment (ZhongRuoyu) [inline .github/workflows/dismiss-approvals.yml:1]: I think this needs a better name, because it's always shown in the list of checks.

```suggestion
name: Check pusher permissions
```

**Comment by p-linnane** [comment]
If we have to hardcode maintainers, it should only be members of the maintainer team, not members.

**Comment by ZhongRuoyu** [comment]
> If we have to hardcode maintainers, it should only be members of the maintainer team, not members.

The membership of `@Homebrew/maintainers` (and other Homebrew org teams) is not public. So I'm playing it safe for now.

**Comment by p-linnane** [comment]
> The membership of `@Homebrew/maintainers` (and other Homebrew org teams) is not public. So I'm playing it safe for now.
> 
The list of maintainers is in the `brew` README, so it should be fine. Better than scoping too wide.

**Review by carlocab — APPROVED** [review]
Members are fine for now; better than nothing.

**Comment by ZhongRuoyu** [comment]
Since it's not hard to list public maintainers... Let me do that anyway. Nothing against non-maintainer members; just being extra cautious.

**Review by p-linnane — APPROVED** [review]
Thanks for the extra diligence with limiting the scope to maintainers!

**Review by carlocab — APPROVED** [review]
Let's merge this for now until we have a better solution. (I'd do it myself but I'm not expecting to be around to fix any potential breakage.)

**Comment by p-linnane** [comment]
I will be around enough to approve a revert if we need it.

**Comment by ZhongRuoyu** [comment]
Sadly, this didn't work on non-Homebrew branches. `${{ secrets.HOMEBREW_GITHUB_PUBLIC_REPO_TOKEN }}` is not available in forks.

**Comment by p-linnane** [comment]
Let's revert and see what other ideas we can come up with.

**Pull request merged**
